File: 2. Structural Design Patterns/Decorator/Kotlin/DecoratorExample.kt
// Decorator Pattern Example in Kotlin
// Notifier: Basic + Email + SMS + Slack decorators

interface Notifier {
    fun send(message: String)
}

class BasicNotifier : Notifier {
    override fun send(message: String) = println("Base notify: $message")
}

open class Decorator(protected val wrap: Notifier) : Notifier {
    override fun send(message: String) = wrap.send(message)
}

class EmailDecorator(w: Notifier) : Decorator(w) {
    override fun send(message: String) {
        println("Email sent: $message")
        super.send(message)
    }
}

class SMSDecorator(w: Notifier) : Decorator(w) {
    override fun send(message: String) {
        println("SMS sent: $message")
        super.send(message)
    }
}

class SlackDecorator(w: Notifier) : Decorator(w) {
    override fun send(message: String) {
        println("Slack message sent: $message")
        super.send(message)
    }
}

fun main() {
    var n: Notifier = BasicNotifier()
    n = EmailDecorator(n)
    n = SMSDecorator(n)
    n = SlackDecorator(n)
    n.send("System maintenance at 2 AM")
}

File: 2. Structural Design Patterns/Decorator/Typescript/DecoratorExample.ts
// Decorator Pattern Example in TypeScript
// Notifier: Basic + Email + SMS + Slack decorators

interface Notifier {
  send(message: string): void;
}

class BasicNotifier implements Notifier {
  send(message: string): void {
    console.log(`Base notify: ${message}`);
  }
}

abstract class Decorator implements Notifier {
  constructor(protected wrap: Notifier) {}

  send(message: string): void {
    this.wrap.send(message);
  }
}

class EmailDecorator extends Decorator {
  send(message: string): void {
    console.log(`Email sent: ${message}`);
    super.send(message);
  }
}

class SMSDecorator extends Decorator {
  send(message: string): void {
    console.log(`SMS sent: ${message}`);
    super.send(message);
  }
}

class SlackDecorator extends Decorator {
  send(message: string): void {
    console.log(`Slack message sent: ${message}`);
    super.send(message);
  }
}

(function main() {
  let n: Notifier = new BasicNotifier();
  n = new EmailDecorator(n);
  n = new SMSDecorator(n);
  n = new SlackDecorator(n);
  n.send('System maintenance at 2 AM');
})();

export {};
